### DISCUSSION TITLE
Push Notifications sent out on every new song from Liquidsoap.

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3866] [General]
Hello everyone.
Is there a way in Liquidsoap to do a Push Notification to a server every time a song changes?

I am using this to write the `Filename` to the file every time a new song starts.
(The old file is deleted, and a new file is created on every new song.)
`file.write(data="#{filename}\n", append=true, "/var/www/stream.example.com/hls/101/101.txt")`

What I am doing with it.

> 1: Get the data from the 101.txt to create a cookie with the `filename` And insert a record into a database
> 2: Run a Refresh on the page, which checks the URL for the `101.txt` to see if it has changed.
> 2a: If it has changed, create a new cookie with the new value.
> 2b: If it has not changed, then do nothing.

The above works well since I am creating the cooking. But, what I would like to do is have a Push Notification sent to the page with the data in the 101.txt file, so all this can be done behind the scenes without using a REFRESH on a page, as it would have to do that refresh all the time, and that is not something that I want to do. Talk about wasted resources.

**HLS**—does have a push, but I think it is strictly for the stream, not to send a notification URL.

**Nginx** (I'm unsure if this is the same thing.)
Has a module: [nginx-push-stream-module](https://github.com/wandenberg/nginx-push-stream-module).

Any suggestions on Push Notifications sent out on every new song?

**Comment by gAlleb** [comment]
Hi @carrzkiss! There are several ways to do it. You can start by reading this thread https://github.com/savonet/liquidsoap/discussions/3782 where with help and guidance from @vitoyucepi different approaches are discussed.

However I ended up using `centrifugo` with `nodejs` server. For your issue I don't think you need a `nodejs` server and can use just Liquidsoap and Centrifugo. Please read about it [here](https://centrifugal.dev/) if you'd like.

It can do lots of stuff but also It will allow you to send LS notifications. On a client side you then can get them via Websocket (unidirectional approach for this case) or Server sent events - so you don't have to write your own WS or SSE servers.

No reloading, no refreshing, no XHR requests - 1 stable connection which lasts.

To start, you have to install centrifugo first. To make it simple go to your user directory (`/home/liquid`) and install it:

```
cd /home/liquid
mkdir centrifugo
cd centrifugo
curl -sSLf https://centrifugal.dev/install.sh | sh
```

Create a systemd file (like you did for liquidsoap, but this time for centrifugo) - `sudo nano /etc/systemd/system/centrifugo.service`  with following contents

```ruby
[Unit]
Description=Centrifugo Server
After=network.target syslog.target

[Service]
LimitNOFILE=30000
ExecStart=/home/liquid/centrifugo/centrifugo --config /home/liquid/centrifugo/config.json
ExecReload=/bin/kill -HUP $MAINPID
ExecStop=/bin/kill -SIGTERM $MAINPID
TimeoutStopSec=5
KillMode=control-group
RestartSec=2
Restart=always
SyslogIdentifier=centrifugo

[Install]
WantedBy=multi-user.target
Alias=centrifugo.service
```

Run `sudo systemctl daemon-reload` and `sudo systemctl enable centrifugo`.

Next return to centrifugo directory and make  `config.json` file (you can use `yaml` or `taml` files as well - but you have tto change systemd config accordingly):

```
cd /home/liquid/centrifugo
./centrifugo genconfig
```

It will generate config file. Edit it (you might need to delete `token_hmac_secret_key` line, but keep admin_password, api_key, etc.) With following:

```ruby
{
  "port": 9998,
  "admin_password": "d0683813-0916-4c49-979f-0e08a686b727",
  "admin_secret": "4e9eafcf-0120-4ddd-b668-8dc40072c78e",
  "api_key": "d7627bb6-2292-4911-82e1-615c0ed3eebb",
  "api_insecure": false,
  "allowed_origins": [
    "*"
  ],
  "admin": true,
  "uni_sse": true,
  "websocket_disable": true,
  "uni_websocket": true,
  "uni_http_stream": true,
  "internal_port": 9998,
  "allow_subscribe_for_client": true,
  "allow_anonymous_connect_without_token": true,
  "admin_insecure": false,
  "namespaces": [
    {
      "name": "station",
      "history_size": 10,
      "history_ttl": "30s",
      "allow_subscribe_for_client": true,
      "allow_subscribe_for_anonymous": true,
      "allow_history_for_client": true,
      "allow_history_for_anonymous": true
    }
  ],
  "uni_sse_handler_prefix": "/connection/sse",
  "uni_websocket_handler_prefix": "/connection/ws"
}

```

And save. If you made no mistakes you can start the service via `sudo systemctl start centrifugo`. It would start on port 9998 and you can access it via yoursite.com:9998. But I suggest you not open this port and instead proxy everything through `nginx`. Just 3 steps:

1. Insert this into your nginx config (`/etc/nginx/nignx.conf` within `http {  }` block)
```ruby
upstream centrifugo {
    server 127.0.0.1:9998;
}

map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      close;
}
```
2. Insert this into you site config ('/etc/nginx/sites-enabled/yourconfig' within server block)
```ruby
location /centrifugo/ {
        rewrite ^/centrifugo/(.*)        /$1 break;
        proxy_pass http://centrifugo;
        proxy_pass_header Server;
        proxy_set_header Host $http_host;
        proxy_redirect off;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Scheme $scheme;
    }

    
location /centrifugo/connection {
        rewrite ^/centrifugo(.*)        $1 break;
        proxy_pass http://centrifugo;
        proxy_buffering off;
        keepalive_timeout 65;
        proxy_read_timeout 60s;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Scheme $scheme;
        proxy_set_header Host $http_host;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;
    }
```
3. Restart nginx server.

Nice! Now you can access you `centrifugo` via `https://yoursite.com/centrifugo` with your password.

<hr>

Now it's time to edit your LS config!

Paste this somewhere after your `music = playlist(mode = 'normal', reload = 1, reload_mode = 'seconds', "https://www.example.com/DJ/101.asp")`

```ruby
def send_mq_event(payload)

  def handle_http_error(error)
    log.important(
      "Unable to send mq event. #{error.message}",
      label="http"
    )
  end
  
  def handle_http_response(result)
    if result.status_code != 200 then
      log.important(
        "Unable to send mq event. #{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="http"
      )
    end
  end

#########centrifugo#############
data = {
  channel = "station:radio",
  data = payload
       }

data = json.stringify(data, compact=true)
auth = "Your API KEY from config.json"
##############################

 response = 
 try
    http.post(
    "http://localhost:9998/api/publish",
    data=data,
    headers=[
      ("Content-Type", "application/json"),
      ("X-API-key", "#{auth}"),
    ],
    )
 catch error do
 handle_http_error(error)
 null()
 end
ignore(null.map(handle_http_response, response))
()
end
```

This function will be used to send data to your centrifugo server.

Next function will collect what you need on track change and send it to your centrifugo server via the above function.

```ruby

def get_send_data(m)

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
np = {
     now_playing = {
        artist = null_list("artist", m)
        title = null_list("title", m)
        filename = null_list("filename", m)
        # etc...
     }
 }

# 
# We don't need to stringify data here (`send_mq_event(json.stringify(np, compact=true))`) - 
# we will do it all together in `send_mq_event` function. This would be easy to parse later. So
# 
send_mq_event(np)
#
# 

end

radio101.on_metadata(get_send_data)
```

This will however really send data on every track change so it's not good enough. To send data every 10 seconds for example you can use this function instead:

```ruby 
def get_send_data(radio101)
 
def write_data()
  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
m = radio101.last_metadata() ?? []

np = {
     now_playing = {
        artist = null_list("artist", m)
        title = null_list("title", m)
        filename = null_list("filename", m)
        # etc...
     }
 }

# 
# We don't need to stringify data here (`send_mq_event(json.stringify(np, compact=true))`) - 
# we will do it all together in `send_mq_event` function. This would be easy to parse later. So
# 
send_mq_event(np)
#
# 

end
thread.run(write_data, every=10.0, fast=false)
end

get_send_data(radio101)
```

We can't control events on connect to WS or SSE with Centrifugo as easily as with self-made WS/SSE server. Here - https://github.com/savonet/liquidsoap/discussions/3782#discussioncomment-8771678 - I've made an example of a workaround using nodejs server with full backend server code which will allow you to get initial data when you load a site and then receive data every 20 seconds + when track changes data will be sent immediately as well.

Great, now your data will be sent to Centrifugo on every track change or every 10 seconds! And then pushed to your site where you can establish WebSocket or SSE (not together) connection with some simple JavaScript functions. No reloading, no refreshing, no XHR requests - 1 stable connection which lasts. All you have to do is parse the json on a client side. Note that this is the simplest use case and you can do much much more. You can send different data to different channels ("station:radio" and "station:radio2"), then push this to your site and parse at the same time. Or you can send data every n seconds to the backend server, manage it in a way you want and send it to centrifugo.

I really encourage you to read the docs.

Example of `js` functions to establish connection to your centriufgo server and start getting updates from liquidsoap:

<details> <summary>Functions</summary> 

```js
let evtSource = null;
let subs = {
    "station:radio": {},
};
const sseUri = "https://yoursite.com/centrifugo/connection/sse?cf_connect="+JSON.stringify({
  "subs": subs
});
function initEventsSSE() {

  if (evtSource === null || evtSource.readyState === 2) {
    evtSource = new EventSource(sseUri);

    evtSource.onerror = (err) => {
      console.error("Now Playing: EventSource failed:", err);
    };

    evtSource.onopen = function() {
      console.log("Now Playing: Server connected.");
    };

    evtSource.onmessage = (event) => {
      const jsonData = JSON.parse(event.data);
   // Do what you want with data
    }
  }
}

// Or WebSocket

function initEventsWS() {
let socket = new WebSocket("wss://yoursite.com/centrifugo/connection/ws");

socket.onopen = function(e) {
  socket.send(JSON.stringify({
    "subs": {
      "station:radio": {},
    }
  })
)};

socket.onmessage = function(e) {
  const jsonData = JSON.parse(e.data);
 // Use your data
};
}

document.addEventListener('readystatechange', event => {
  if (event.target.readyState === "complete") {
    initEventsSSE();
    // or
    //initEventsWS();
  }
});

```
</details>

**Comment by carrzkiss** [comment]
Thank you, Stefan.
Where would the URL be for accepting the PUSH from LS and sending it to the server for processing?
(OR)
Is this to send a push to the centrifugo?

**Reply by gAlleb** [reply]
Liquidsoap sends data to http://localhost:9998. Then Centrifugo receives it  awaits connection from a client to push data further. It's all above!

**Reply by carrzkiss** [reply]
Will that be these two sections?
```
const sseUri = "https://yoursite.com/centrifugo/connection/sse?cf_connect="+JSON.stringify({
  "subs": subs
});
let socket = new WebSocket("wss://yoursite.com/centrifugo/connection/ws");
```
If so, what do I need to do on my server to allow this connection?

**Reply by gAlleb** [reply]
They are allowed since you've proxied them with nginx on default web ports.

**Reply by carrzkiss** [reply]
I am completely lost on this one.
OK. The URL to the server where the updates to the database happen is like this.
`https://www.example.com/process/file.asp?StationID=101`
Where would that be placed within the code so it knows where to send the PUSH?
I do not see anywhere in the code where this URL will be placed.
All I see is **LS** talking with **Centrifugo**

**Reply by gAlleb** [reply]
Could you tell me what for do you need these notifications?  What is the end purpose?

**Reply by carrzkiss** [reply]
OK, I am sending a push to the server with the filename.
On the server, it takes the filename and creates a record every time it is played.
The records are then tallied so I can display on the song page the number of times this song has been played.
When a person clicks on the `Played 80 times` link. (Example, 80 times it has been played.)
They will be taken to a page where they can see what channels have played the song and on what day and time it was played on that channel.
It is a tracking system.

I have already written a PUSH in Powershell that works well, but the way it is written, if I had a hundred channels, it would kill the server it is running on. As of right now, it checks every second. I am going to see if I can change it.
I was hoping I could do a push from LS instead, as it will only do a check every time the song changes.

It would be cool if there were a way in the LS script to have something like this.
This first code will write the filename (the Actual filename) to the 101.txt file. 
(This is what I have in place to get the information to send to the server).
```
filename = m["initial_uri"]
file.write(data="#{filename}\n", append=true, "/var/www/stream.example.com/hls/101/101.txt")
```
This here is a sample of Powershell
`$Response = Invoke-WebRequest -URI https://demo.example.com/test/getFile.asp?FileName=$($gethtml)`
Changing up a little for LS (The ending)
`Response = Invoke-WebRequest -URI https://demo.example.com/test/getFile.asp?FileName="#{filename}`

Having something like this invoked into LS would make it extremely powerful.
Every time a new song is played, this will be triggered, and the information will be updated on the admin's servers to their heart's content.

So, that is what I am doing and what I would LOVE to see in LS one day. Unless it is already there, but I just have not found it yet.

**Reply by gAlleb** [reply]
Then it should be easier and done with one function (see below). 

I am not familiar with `asp` methods. But if it has `getfile` maybe it has something else you can use to pass a string instead and then translate your `Invoke-WebRequest` into `curl` and `http.post` method. Then you should be able to use it with the function below 

```ruby

def get_send_data(m)
 
  def handle_http_error(error)
    log.important(
      "Unable to send mq event. #{error.message}",
      label="http"
    )
  end
  
  def handle_http_response(result)
    if result.status_code != 200 then
      log.important(
        "Unable to send mq event. #{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="http"
      )
    end
  end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
filename = null_list("initial_uri", m)

 response = 
 try
# 
# Here you should construct your request with `filename`
# 
    http.post(
    "https://demo.example.com/....",
    data=filename,
    headers=[
      # Headers here
    ],
    )
# 
# 
# 
 catch error do
 handle_http_error(error)
 null()
 end
ignore(null.map(handle_http_response, response))
()
end

radio101.on_metadata(get_send_data)
```

or use pure `curl` instead of `http.post` block above. 
For example to make it more clear - if I want to use curl to send something to Telegram Bot - I could use this instead of http.post block above:
```
curlcmdUtils = "curl -s --data 'text=#{filename}' --data 'chat_id=#{CHATID}' 'https://api.telegram.org/bot'#{BOTID}'/sendMessage'"
cmdUtils = "#{curlcmdUtils}"
process.run(cmdUtils)
```

If I want to send something to my API I can use http.post

```
    http.post(
    "http://localhost/api",
    data=data,
    headers=[
    # headers
    ],
    )

```

I'm still unsure what you are doing)) But to make it work you should translate your `Invoke-WebRequest` into `curl` and/or `http.post` and maybe use another asp method on server side since we are not fetching file anymore and can send just string.  How exactly we can do it - I have no idea.

**Reply by carrzkiss** [reply]
As for the asp side of things, it does not really matter. As long as I can send a url with a QueryString attached to it with the filename, all is good.

I will test this out tomorrow.
Thank you.

**Reply by gAlleb** [reply]
Then maybe `http.get("https://demo.example.com/test/getFile.asp?FileName="#{filename}")` or `http.post("https://demo.example.com/test/getFile.asp?FileName="#{filename}")` would be enough inside that block

**Reply by Moonbase59** [reply]
Just remember query strings also have to be URL-encoded, or you’ll run into the same problem as in the other thread.

**Reply by carrzkiss** [reply]
The Querystring does not need to be URL-Encoded; it is nothing but numbers and the extension.
02191971115908.mp3
The files are created by date and time.
The example is my birthday and hour.

**Reply by carrzkiss** [reply]
Testing---
Code 1
```
    http.post("https://demo.example.com/test/getFile.asp?FileName=",
    data=filename,
    headers=[
      # Headers here
    ],
    )
```
Code 1 Error
```
Error 5: this value has type
data=filename,
{string}
but it should be a subtype of
 _? (inferred at /home/liquid/liquidsoap/101.liq, line 62, char 4-64)
```

Just this and all the above removed.
Code 2
`http.post("https://demo.example.com/test/getFile.asp?FileName="#{filename}")`
Code 2 Error
```
Error 2: Parse error
catch error do
```

It seems that it is not getting the filename.
`filename = null_list("initial_uri", m)`

I tried this, but it did not work either. (From the HLS script)
Code 3
```
  data = {
    now_playing = {
      filename = null_list("initial_uri", m)
     }
  }
```
I get this error with code 3 with `Code 1`
```
Error 4: Undefined variable filename
data=filename,
```
I get this error with code 3 with `Code 2`
(Same with or without the changed filename code.)
```
Error 2: Parse error
catch error do
```

removing the double quote before the variable 
`http.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")`
I receive the same error as above, 
```
Error 4: Undefined variable filename
 http.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")
```

I changed 
`filename = null_list("initial_uri", m)`
to
`filename = m["initial_uri"]`
And I get an error
```
Error 5: this value has type
radio101 = metadata.map(update=false, hls_meta,get_send_data,music)
At /home/liquid/liquidsoap/101.liq, line 113, char 48-61:
source(_)
 but it should be a subtype of
 (_) -> _ (inferred at /home/liquid/liquidsoap/101.liq, line 42 char >
```

And changing back to using
`filename = null_list("initial_uri", m)`
I now get the following error.

```
Error 6: Cannot apply that parameter because the function at /home/liquid/liquidsoap/101.liq, line 109, char 12-68
radio101 = metadata.map(update=false, hls_meta,music,get_send_data)
At /home/liquid/liquidsoap/101.liq, line 109, char 54-67:
has no more unlabeled argument!
```
OK, that is enough for one morning.

(I removed my last two posts.
Reason: I had removed, without knowing, `,get_send_data` from the radio101.
So, it was not running the script.
So, everything above was tested. They are the accurate results.)

**Reply by gAlleb** [reply]
Don't do this
```
radio101 = metadata.map(update=false, hls_meta,music,get_send_data)
```
<hr>
Just paste this somewhere after radio101 var is defined.

```ruby
def get_send_data(m)
 
  def handle_http_error(error)
    log.important(
      "Unable to send mq event. #{error.message}",
      label="http"
    )
  end
  
  def handle_http_response(result)
    if result.status_code != 200 then
      log.important(
        "Unable to send mq event. #{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="http"
      )
    end
  end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
filename = null_list("initial_uri", m)

 response = 
 try

   http.get("https://demo.example.com/test/getFile.asp?FileName=#{filename}")
 
 catch error do
 handle_http_error(error)
 null()
 end
ignore(null.map(handle_http_response, response))
()
end

radio101.on_metadata(get_send_data)
```

or you can paste it just after your `music = playlist()` but then change `radio101.on_metadata(get_send_data)` to `music.on_metadata(get_send_data)`

**Reply by carrzkiss** [reply]
I placed it on top and am using `music.on_metadata(get_send_data)`
It runs and plays but will not send Push/Post to the server.

I also tried this as well. It plays but will not post.

```
    https.post("https://demo.example.com/test/getFile.asp?FileName=",
    data=filename,
    headers=[
      # Headers here
    ],
    )
```

I also changed the
`label="http"`
to
`label="https"`
In both places to see if that would help, but still no post.

**Reply by carrzkiss** [reply]
2 things.
1st: I just noticed you had `get` and not `post` so I fixed that.
I also noticed I had an error in the code, a `comma`, but LS did not pick up on it, which means that LS is **not** registering this block of code, and LS is not using it.

I have it under `music = playlist()`
And I have this in place.
`music.on_metadata(get_send_data)`
And on further down the code, I have this.
`radio101 = metadata.map(update=false, hls_meta,music)`

Something is a miss, and it is not picking up on the `get_send_data`

I just changed it to this, placing it right under, but still no post.
```
radio101 = metadata.map(update=false, hls_meta,music)
radio101.on_metadata(get_send_data
```

**Reply by gAlleb** [reply]
Well, you've said that

> As for the asp side of things, it does not really matter. As long as I can send a url with a QueryString attached to it with the filename, all is good.

The function should call an url: `https://demo.example.com/test/getFile.asp?FileName=initial_uri".

You can comment out here and then see the logs

```ruby
def handle_http_response(result)
   # if result.status_code != 200 then
      log.important(
        "#{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="http"
      )
   #  end
  end
```

**Reply by carrzkiss** [reply]
```
 Warning 4: Unused variable null_list
end
 list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
def null_list(key, _list)
```

On this
```
  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
```

**Reply by carrzkiss** [reply]
I removed it, and it still runs, but no post and no more errors.
IS there a way to see the output of the http.post?

**Reply by gAlleb** [reply]
comment out just 2 lines

```ruby
def get_send_data(m)
 
  def handle_http_error(error)
    log.important(
      "Unable to send mq event. #{error.message}",
      label="http"
    )
  end
  
  def handle_http_response(result)
# 
# 
# 
# 
 #    if result.status_code != 200 then
      log.important(
        "#{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="http"
      )
#  end
# 
# 
# 
# 
  end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
filename = null_list("initial_uri", m)

 response = 
 try

   http.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")
 
 catch error do
 handle_http_error(error)
 null()
 end
ignore(null.map(handle_http_response, response))
()
end

radio101.on_metadata(get_send_data)

```

and look at the logs

We are doing it this so you could see the output

**Reply by carrzkiss** [reply]
Hang on.
What you said makes me think of something.
I will post back later, I think I might have just figured it out, I hope.

**Reply by carrzkiss** [reply]
OK, so what you wrote made me think.
>> `The function should call a url: https://demo.example.com/test/getFile.asp?FileName=initial_uri".`
The `initial_uri` gets the file path. That is one part that I totally forgot about.
I can remove every before the last forward slash `/` on the server side, but is there a way to do it in liquidsoap?
Song string example
`https://demo.example.com/host/Files/BandName/AlbumName/02082024142522311.mp3`
To only get `02082024142522311.mp3`

**Reply by carrzkiss** [reply]
Also, I copied what you posted; just removed my code, and replaced it with what you provided above. The output from what you provided is what I posted as the error output.

This is mine, the same as what you first posted.

```
def handle_http_response(result)
   # if result.status_code != 200 then
      log.important(
        "#{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="https"
      )
   #  end
  end
```
I copied yours because I was confused as your original code was this.
`"Unable to send mq event. #{result.status_code} #{result.status_message} \`
Without the leading `"#`

**Reply by gAlleb** [reply]
> Also, I copied what you posted; just removed my code, and replaced it with what you provided above. The output from what you provided is what I posted as the error output.

No idea. I've copied it too and tested and it works just fine and gives me logs.

```ruby
2024/04/10 19:41:04 [http.get:3] Requested url "https://demo.example.com/test/getFile.asp?FileName=/Users/liquidsoap/music/05. Dance Me to the End of Love [Live].mp3" different from normalized url: "https://demo.example.com/test/getFile.asp?FileName=/Users/liquidsoap/music/05.%20Dance%20Me%20to%20the%20End%20of%20Love%20%5BLive%5D.mp3". Either fix it or use `normalize_url=false` to disable url normalization!
```

So you should check for mistakes.

> but is there a way to do it in liquidsoap?

Sure there is. But I don't know how.

**Reply by carrzkiss** [reply]
OK, the Push is working.
I just ran a test on the server, and the database got an EMPTY record, BUT it got it.
So, that means that it does not have a value. The record is empty.
This means this is not passing the value to the data portion of the https.post.
`filename = m["initial_uri"]`

`filename = null_list("initial_uri", m)` (This does not work.) 
```
Error 5: 
this value has type, 
data=filename, {string}
```

**Reply by gAlleb** [reply]
I dont see `data=filename` in here https://github.com/savonet/liquidsoap/discussions/3866#discussioncomment-9073454

I am not familiar with asp methods. Maybe you should translate your Invoke-WebRequest into curl and http.post method with headers and data attribute. Not just a reqeust to an url. I don't know what server is expecting.

**Reply by carrzkiss** [reply]
I just added a value to the `FileName=ThisIsNotWorking`
And that was inserted into the database.
Just need to find out WHY the filename is not passing.

**Reply by carrzkiss** [reply]
This is my code for that area.

```

filename = m["initial_uri"]
 response =
 try
    https.post("https://demo.example.com/test/getFile.asp?FileName=",
    data=filename,
    headers=[
      # Headers here
    ],
    )

```

**Reply by carrzkiss** [reply]
OK, I will try with just the string again and see what happens.

**Reply by carrzkiss** [reply]
WORKING!!!!
OK, and the string can be handled on my end.
So, it seems it will only work with the 
`https.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")`

**Reply by gAlleb** [reply]
Thank God!

**Reply by carrzkiss** [reply]
I was trying so many different things, and what got me all messed up is why I removed the `https.post` string. There was double the code in my liq file that I did not know got left behind.
That was causing some issues, and once I found it and removed it, I was already messing with the other code and had removed the string.

I wish I could accept your post as an answer. [9071731](https://github.com/savonet/liquidsoap/discussions/3866#discussioncomment-9071731)

Thank you so much. I know that was frustrating, but thank you for sticking with me here.
I think this should have me tied over for a while.
I am able to send data LIVE on every new song without wasting resources on either server, which is wonderful.

Thank you, Stefan, you rock, my friend.
Once the site is live, I will come in and create a new thread to let everyone know.
The Radio part of it will be limited to listeners for a while, but it will be available once the site is live.

Thank you.

**Comment by carrzkiss** [comment]
Here is the working script for those who are thinking of running a similar setup.
100% of the credit goes to [Stefan](https://github.com/gAlleb) and [Vito](https://github.com/vitoyucepi).
Luckily, I created a script to write to a text file the other day, which helped solve the code-breaking issue. 
`filename = null_list("initial_uri", m)`
Line 24, replace with `filename = m["initial_uri"]`
```
==========MetaData Information Here=================#
def get_send_data(m)

  def handle_http_error(error)
    log.important(
      "Unable to send mq event. #{error.message}",
      label="https"
    )
  end

def handle_http_response(result)
   # if result.status_code != 200 then
      log.important(
        "#{result.status_code} #{result.status_message} \
         #{string.quote(result)}",
        label="https"
      )
   #  end
  end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
filename = m["initial_uri"]
 response =
 try
   https.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")
 catch error do
 handle_http_error(error)
 null()
 end
ignore(null.map(handle_http_response, response))
()
end

#music.on_metadata(get_send_data)
radio101.on_metadata(get_send_data)
#==========MetaData Information Here=================#
```
Enjoy.
Also, add it as an include.
I am going to do that right now. The include helps keep everything organized and neat.
And it keeps your main liq file clean and tidy.
Create a folder called `plugins` and add all your scripts in there.
Copy the code from your `main.liq` file, paste it into a new file, and name it.
Then, place this in the same area where the code was at.
`%include "plugins/PUSH.liq"`

Good Luck.
Wayne

**Reply by gAlleb** [reply]
I'm not sure why `filename = null_list("initial_uri", m)` doesn't work. it should

We use this function to secure ourselves and return null in case something goes wrong. Also since you are using `https.post("https://demo.example.com/test/getFile.asp?FileName=#{filename}")` again, you could try one more time to see if it works.

```ruby
  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end
 
filename = null_list("initial_uri", m)
```

If you will use `filename = m["initial_uri"]` instead of `filename = null_list("initial_uri", m)` then you don't need the above `null_list(key, _list)` function.

p.s. Also credit should go to @vitoyucepi as he is the initial mind behind these functions.

**Reply by carrzkiss** [reply]
Updated to credit both.
OK, that is weird. It is working now. But why would it not work before?

**Reply by gAlleb** [reply]
Because of this `data=filename` in one of your iterations.

**Reply by carrzkiss** [reply]
Resolved the issue.
I had to take a look at the original code that was working when this was done last year, and was able to figure out what was missing.
So, now I have a template that I will use to create new stations.
Glad to have this part taken care of.

Have a good one, Stefan.
Wayne
